### ISSUE TITLE
make:entity wrong setter for non nullable fields

### ISSUE THREAD

**Issue by sbruev** [symfony/maker-bundle#198]
Hi!
It's a minor syntax mistake, but right now generated setters allow to pass the null value to non nullable fields, if they are ManyToOne relationship.
```php
    /**
     * @ORM\ManyToOne(targetEntity="App\Entity\AnotherEntity")
     * @ORM\JoinColumn(nullable=false)
     */
    private $anotherEntity;

    //...

    public function setAnotherEntity(?AnotherEntity $anotherEntity): self
    {
        $this->anotherEntity = $anotherEntity;

        return $this;
    }
```
The arg should be `AnotherEntity` not `?AnotherEntity`

Cheers!

**Comment by lenar** [comment]
Are you sure this is a bug? This would allow to use validations, otherwise the failure occurs too early.
I would say that currently there is a bug with ordinary fields which don't have ? before type if nullable=false. Failure occurs too early, makes it hard to validate. Maybe it should be configurable if one wants to allow nulls or not for setters.

**Comment by weaverryan** [comment]
Hmm, interesting. Are you using the form submit, and it’s submitting the values as null?

**Comment by lenar** [comment]
No, not using form submit. "Legacy" code, build with older Symfony releases. Manual validations with Symfony validation/constraint classes.

Anyway it should be selectable (by argument to console command) if you want very strict definitions (nulls not allowed if nullable=false and thus fail with PHP error) or you want to handle things yourself later with validations (and this seems to be the default with older releases and doctrine:generate:entitites).

And I really prefer handling nulls with validations and not manually modifying string => ?string in setters of generated entities.

**Comment by lenar** [comment]
Looking at older (doctrine:generate) generated entities, the setters are `setSomething($something = null)` irrelevant of `nullable` of column annotation. They were generated automatically. And it makes sense IMHO.

**Comment by javiereguiluz** [comment]
Even if this is correct in theory ... I'm afraid that in practice, when using Symfony Forms, is not correct. You must make everything nullable or the app won't work. That's why we did that in the Symfony Demo app for example: https://github.com/symfony/demo/blob/master/src/Entity/Post.php

**Comment by mesolaries** [comment]
+1 for this to be non-nullable. 
One thing to consider is the generated `remove*` methods for bidirectional associations, which set the owning side to `null`.

Also, when using typed properties, phpstan/phpstan-doctrine reports a `doctrine.associationType` error:

```php
    /**
     * @ORM\ManyToOne(targetEntity="App\Entity\AnotherEntity")
     * @ORM\JoinColumn(nullable=false)
     */
    private ?AnotherEntity $anotherEntity = null;

    //...

    public function setAnotherEntity(?AnotherEntity $anotherEntity): self
    {
        $this->anotherEntity = $anotherEntity;

        return $this;
    }
```

`Property Entity::$anotherEntity type mapping mismatch: property can contain AnotherEntity|null but database expects AnotherEntity`

**Comment by Amoifr** [comment]
I gave this a shot in #1797 — as a concrete design proposal rather than a "fix everything" change. 🙂

The thread stalled on a genuine constraint: the setter has to stay nullable for **bidirectional** relations, since the inverse-side `remove*()` sets the owning side back to null (orphanRemoval). So #1797 only tightens the **provably-safe subset** — a non-nullable, **unidirectional** `ManyToOne` — which has no such code path:

```php
public function setCategory(Category $category): static  // was: ?Category
```

Bidirectional and nullable relations stay exactly as they are. It mirrors what non-nullable `OneToOne` setters already do today.

I kept it **setter-only** on purpose. Making the backing *property* non-nullable too (which would also address the phpstan-doctrine `doctrine.associationType` point @mesolaries raised) reopens the Forms / uninitialized-property concern @javiereguiluz mentioned, so that feels like a separate discussion. Happy to go there if there's appetite — feedback very welcome! 🙏
